File: swift/tests/WalletKitTests/DeviceKeystoreTests.swift
import CryptoKit
import Foundation
import Security
import XCTest
@testable import WalletKit

final class DeviceKeystoreTests: XCTestCase {
    private let account = "test-account"

    func testSealAndOpenRoundTrip() throws {
        let service = uniqueKeystoreService()
        defer { deleteKeychainItem(service: service, account: account) }

        let keystore = IOSDeviceKeystore(service: service, account: account)
        let associatedData = Data("ad".utf8)
        let plaintext = Data("hello".utf8)

        let ciphertext = try keystore.seal(
            associatedData: associatedData,
            plaintext: plaintext
        )
        let opened = try keystore.openSealed(
            associatedData: associatedData,
            ciphertext: ciphertext
        )

        XCTAssertEqual(opened, plaintext)
    }

    func testAssociatedDataMismatchFails() throws {
        let service = uniqueKeystoreService()
        defer { deleteKeychainItem(service: service, account: account) }

        let keystore = IOSDeviceKeystore(service: service, account: account)
        let plaintext = Data("secret".utf8)

        let ciphertext = try keystore.seal(
            associatedData: Data("ad-1".utf8),
            plaintext: plaintext
        )

        XCTAssertThrowsError(
            try keystore.openSealed(
                associatedData: Data("ad-2".utf8),
                ciphertext: ciphertext
            )
        )
    }
}

File: kotlin/walletkit-tests/src/test/kotlin/org/world/walletkit/DeviceKeystoreTests.kt
package org.world.walletkit

import kotlin.test.Test
import kotlin.test.assertFails
import kotlin.test.assertTrue

class DeviceKeystoreTests {
    @Test
    fun sealAndOpenRoundTrip() {
        val keystore = InMemoryDeviceKeystore()
        val associatedData = "ad".encodeToByteArray()
        val plaintext = "hello".encodeToByteArray()

        val ciphertext = keystore.seal(associatedData, plaintext)
        val opened = keystore.openSealed(associatedData, ciphertext)

        assertTrue(opened.contentEquals(plaintext))
    }

    @Test
    fun associatedDataMismatchFails() {
        val keystore = InMemoryDeviceKeystore()
        val plaintext = "secret".encodeToByteArray()
        val ciphertext = keystore.seal("ad-1".encodeToByteArray(), plaintext)

        assertFails {
            keystore.openSealed("ad-2".encodeToByteArray(), ciphertext)
        }
    }

    @Test
    fun reopenWithSameKeyMaterialCanOpenCiphertext() {
        val keyBytes = randomKeystoreKeyBytes()
        val firstKeystore = InMemoryDeviceKeystore(keyBytes)
        val secondKeystore = InMemoryDeviceKeystore(keyBytes)
        val associatedData = "ad".encodeToByteArray()
        val plaintext = "hello".encodeToByteArray()

        val ciphertext = firstKeystore.seal(associatedData, plaintext)
        val opened = secondKeystore.openSealed(associatedData, ciphertext)

        assertTrue(opened.contentEquals(plaintext))
    }
}
